### PR TITLE
Improve wps outputs

### PR DESCRIPTION
**Pull request by ccmoloney**
Fixing up the WPS outputs - including output links in the status document when successful.

### CODE REVIEW

File: request-handler/src/main/java/au/org/aodn/aws/wps/operation/ExecuteOperation.java [hunk 1]
```diff
@@ -143,4 +144,72 @@ public void validate(Properties config) throws ValidationException
         //  Validate execute operation
     }
 
+    private void debugListJobs(String awsRegion, String queueName)
```
Comment (jonescc) [inline request-handler/src/main/java/au/org/aodn/aws/wps/operation/ExecuteOperation.java]: Remove?  Can't determine queue position with current AWS batch implementation

File: wps-common/src/main/java/au/org/aodn/aws/wps/status/StatusHelper.java [hunk 2]
```diff
@@ -1,20 +1,179 @@
 package au.org.aodn.aws.wps.status;
 
+import net.opengis.ows._1.*;
+
+import net.opengis.wps._1_0.*;
+import net.opengis.ows._1.ObjectFactory;
```
Comment (jonescc) [inline wps-common/src/main/java/au/org/aodn/aws/wps/status/StatusHelper.java]: Is this the wrong import?  Doesn't seem to be used below and references to net.opengis.wps._1_0.ObjectFactory are fully qualified.

File: wps-common/src/main/java/au/org/aodn/aws/wps/status/StatusHelper.java [hunk 3]
```diff
@@ -1,20 +1,179 @@
...
+        outputIdentifierCode.setValue(outputIdentifier);
+        output.setIdentifier(outputIdentifierCode);
+        if(response.getProcessOutputs() == null) {
+            ExecuteResponse.ProcessOutputs outputs = objectFactory.createExecuteResponseProcessOutputs();
+            response.setProcessOutputs(outputs);
+        }
+        response.getProcessOutputs().getOutput().add(output);
+    }
+
+    public static void addExecuteLiteralProcessOutput(ExecuteResponse response, String outputIdentifier, InputStream inStream)
```
Comment (jonescc) [inline wps-common/src/main/java/au/org/aodn/aws/wps/status/StatusHelper.java]: Not used anywhere. Would it be better to remove this until we need to support LiteralProcessOutputs?

File: wps-common/src/main/java/au/org/aodn/aws/wps/status/ExecuteStatusBuilder.java [hunk 4]
```diff
@@ -36,71 +26,77 @@ public String getStatusLocation() {
...
-            report.getException().add(type);
-            failed.setExceptionReport(report);
+            //  If outputs were passed - add them to the response
+            if(outputs != null)
+            {
+                for(String currentKey : outputs.keySet())
+                {
+                    String href = outputs.get(currentKey);
+                    LOGGER.info("OUTPUT [" + currentKey + "]=[" + href + "]");
+                    StatusHelper.addExecuteOutputReference(response, currentKey, href);
```
Comment (akashisama) [inline wps-common/src/main/java/au/org/aodn/aws/wps/status/ExecuteStatusBuilder.java]: I think renaming the method "addOutputToResponse" would make the code more readable here.

File: wps-common/src/main/java/au/org/aodn/aws/wps/status/StatusHelper.java [hunk 5]
```diff
@@ -1,20 +1,179 @@
...
 
-    public static final String getStatusDocument(String s3Bucket, String statusFilename, EnumOperation operation, String jobId, EnumStatus jobStatus, String message, String messageCode)
+    static int CHUNK_SIZE = 512;
+
+    public static final String getStatusDocument(String s3Bucket, String statusFilename, String jobId, EnumStatus jobStatus, String message, String messageCode, HashMap<String, String> outputsHrefs)
     {
-        String statusDocument = null;
+        String statusLocation = getS3ExternalURL(s3Bucket, jobId + "/" + statusFilename);
+        ExecuteStatusBuilder statusBuilder = new ExecuteStatusBuilder(statusLocation, jobId);
+        return statusBuilder.createResponseDocument(jobStatus, message, messageCode, outputsHrefs);
```
Comment (jonescc) [inline wps-common/src/main/java/au/org/aodn/aws/wps/status/StatusHelper.java]: This is a bit convoluted - statushelper calls status builder calls status helper to build the document.  Perhaps move to StatusBuilder.

File: wps-common/src/main/java/au/org/aodn/aws/wps/status/ExecuteStatusBuilder.java [hunk 6]
```diff
@@ -36,71 +26,77 @@ public String getStatusLocation() {
...
-            ExceptionReport report = new ExceptionReport();
-            ExceptionType type = new ExceptionType();
-            type.getExceptionText().add(failedMessage);
-            type.setExceptionCode(failedCode);
-            report.getException().add(type);
-            failed.setExceptionReport(report);
+            //  If outputs were passed - add them to the response
+            if(outputs != null)
+            {
+                for(String currentKey : outputs.keySet())
```
Comment (akashisama) [inline wps-common/src/main/java/au/org/aodn/aws/wps/status/ExecuteStatusBuilder.java]: According to the interwebs this is inefficient and you should iterate via the entrySet() method if you want both key and value. Might be more readable as well?

File: wps-common/src/main/java/au/org/aodn/aws/wps/status/StatusHelper.java (hunk 3 above)
Comment (akashisama) [inline wps-common/src/main/java/au/org/aodn/aws/wps/status/StatusHelper.java]: I feel like these two methods are doing such similar things that they should share the same name, eg "addOutputToResponse"

File: wps-common/src/main/java/au/org/aodn/aws/wps/status/StatusHelper.java [hunk 7]
```diff
@@ -1,20 +1,179 @@
...
+        currentTime.setTime(new Date());
+        currentTime.setTimeZone(TimeZone.getTimeZone("GMT"));
+        return DatatypeFactory.newInstance().newXMLGregorianCalendar(currentTime);
+    }
 
-        return statusDocument;
+
+    public static void addExecuteOutputReference(ExecuteResponse response, String outputIdentifier, String outputHref)
+    {
+        net.opengis.wps._1_0.ObjectFactory objectFactory = new net.opengis.wps._1_0.ObjectFactory();
```
Comment (akashisama) [inline wps-common/src/main/java/au/org/aodn/aws/wps/status/StatusHelper.java]: Is the factory necessary? Would be cleaner to just say ```new OutputDataType()``` no?

File: wps-common/src/main/java/au/org/aodn/aws/wps/status/StatusHelper.java [hunk 8]
```diff
@@ -1,20 +1,179 @@
...
+    {
+        GregorianCalendar currentTime = new GregorianCalendar();
+        currentTime.setTime(new Date());
+        currentTime.setTimeZone(TimeZone.getTimeZone("GMT"));
+        return DatatypeFactory.newInstance().newXMLGregorianCalendar(currentTime);
+    }
 
-        return statusDocument;
+
+    public static void addExecuteOutputReference(ExecuteResponse response, String outputIdentifier, String outputHref)
```
Comment (akashisama) [inline wps-common/src/main/java/au/org/aodn/aws/wps/status/StatusHelper.java]: Will this be useful outside of execute status documents? If not it should be moved into the ExecuteStatusBuilder class as a private method.

File: wps-common/src/main/java/au/org/aodn/aws/wps/status/StatusHelper.java (hunk 3 above)
Comment (ccmoloney) [reply wps-common/src/main/java/au/org/aodn/aws/wps/status/StatusHelper.java]: Possibly  :)  It may be useful to someone some day though  :)

**Comment by jonescc** [comment]
Perhaps we should split the builder up according to the type of document to be built e.g. FailedStatusBuilder, SucceededStatusBuilder, AcceptedStatusBuilder etc with common code shared on parent StatusBuilder class.

File: wps-common/src/main/java/au/org/aodn/aws/wps/status/StatusHelper.java (hunk 2 above)
Comment (ccmoloney) [reply wps-common/src/main/java/au/org/aodn/aws/wps/status/StatusHelper.java]: removed.

File: wps-common/src/main/java/au/org/aodn/aws/wps/status/ExecuteStatusBuilder.java (hunk 4 above)
Comment (ccmoloney) [reply wps-common/src/main/java/au/org/aodn/aws/wps/status/ExecuteStatusBuilder.java]: Changed.

File: wps-common/src/main/java/au/org/aodn/aws/wps/status/StatusHelper.java (hunk 7 above)
Comment (ccmoloney) [reply wps-common/src/main/java/au/org/aodn/aws/wps/status/StatusHelper.java]: Changed.

File: wps-common/src/main/java/au/org/aodn/aws/wps/status/StatusHelper.java (hunk 3 above)
Comment (ccmoloney) [reply wps-common/src/main/java/au/org/aodn/aws/wps/status/StatusHelper.java]: Changed.

File: request-handler/src/main/java/au/org/aodn/aws/wps/operation/ExecuteOperation.java (hunk 1 above)
Comment (ccmoloney) [reply request-handler/src/main/java/au/org/aodn/aws/wps/operation/ExecuteOperation.java]: Removed.

File: wps-common/src/main/java/au/org/aodn/aws/wps/status/StatusHelper.java (hunk 5 above)
Comment (ccmoloney) [reply wps-common/src/main/java/au/org/aodn/aws/wps/status/StatusHelper.java]: Changed so the call to build the document is a call to ExecuteStatusBuilder - which then uses StatusHelper.

File: wps-common/src/main/java/au/org/aodn/aws/wps/status/ExecuteStatusBuilder.java [hunk 9]
```diff
@@ -101,6 +118,79 @@ public static void main(String[] args) {
...
+    public static ProcessStartedType getProcessStartedType(String message, Integer percentComplete)
+    {
+        ProcessStartedType started = new ProcessStartedType();
+        started.setValue(message);
+        started.setPercentCompleted(percentComplete);
+        return started;
+    }
+
+
+    public static void addOutputToResponse(ExecuteResponse response, String outputIdentifier, String outputHref)
```
Comment (akashisama) [inline wps-common/src/main/java/au/org/aodn/aws/wps/status/ExecuteStatusBuilder.java]: Could these two methods be private? ExecuteResponse and other jaxb classes should be left as behind the scenes implementation details.

File: wps-common/src/main/java/au/org/aodn/aws/wps/status/ExecuteStatusBuilder.java (hunk 6 above)
Comment (ccmoloney) [reply wps-common/src/main/java/au/org/aodn/aws/wps/status/ExecuteStatusBuilder.java]: Changed to use entrySet()

File: wps-common/src/main/java/au/org/aodn/aws/wps/status/ExecuteStatusBuilder.java (hunk 9 above)
Comment (ccmoloney) [reply wps-common/src/main/java/au/org/aodn/aws/wps/status/ExecuteStatusBuilder.java]: Changed to private.

**Pull request merged**